src/libraries/System.Diagnostics.DiagnosticSource/src/Properties/InternalsVisibleTo.cs

```diff
@@ -0,0 +1,6 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("System.Diagnostics.DiagnosticSource.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb")]
```

Comment:
>InternalsVisibleTo [](http://example.com/codeflow?start=11&length=18)

I really don't like having this attribute. Is there any other way we can avoid that (using reflection for instance?

Reply:
If that's not possible (for whatever reason) please prefer setting the InternalsVisibleTo as an item in the project file instead.

Reply:
Why do you prefer reflection to this attribute Tarek? At present there are ~30 tests that are unit testing internal APIs. I assume with reflection we could get around it, but its not clear what value is to add all the extra code?

Reply:
In general, it creates hard dependencies. at any time, we need to use the same logic in other tests or decided for some reasons changing the test assembly names, or even the public key (in extreme cases), this will be broken. I know reflection is more involving but more flexible.

Reply:
> at any time, we need to use the same logic in other tests

I'm not sure I followed this part of the explanation. Did you mean we decide that we want to create a 2nd test assembly that has similar code in it? I'm not sure what would ever cause us to do that so it feels like an unlikely eventuality to me, but we could always add a 2nd InternalsVisibleTo attribute in such a case.

> decided for some reasons changing the test assembly names, or even the public key

If we changed the test assembly names/keys wouldn't we update the InternalsVisibleTo attribute at the same time? This one also seems like a relatively unlikely eventuality?

> I know reflection is more involving but more flexible.

I'm not sure I am understanding the full value of the flexibility yet. If the only risk we avoid is that in the future we would change/add an InternalsVisibleTo attribute that feels like the benefit is much smaller than the cost.

Reply:
> I'm not sure I followed this part of the explanation. Did you mean we decide that we want to create a 2nd test assembly that has similar code in it? I'm not sure what would ever cause us to do that so it feels like an unlikely eventuality to me, 

We run into some cases in other areas. For example, we wanted to run some standalone tests under specific environments (like with environment variables). It is testing the same area but under some different environment. 
by the way, did you try to run any test case under remote execution and findout if this work fine there?

> but we could always add a 2nd InternalsVisibleTo attribute in such a case.
> If we changed the test assembly names/keys wouldn't we update the InternalsVisibleTo attribute at the same time?

Why we'll need to touch the source assembly to just update the tests? this is exactly the point I am trying avoid :-)

> I'm not sure I am understanding the full value of the flexibility yet. If the only risk we avoid is that in the future we would change/add an InternalsVisibleTo attribute that feels like the benefit is much smaller than the cost.

The flexibility is we can change the test the way we like without ever touching the source assembly. tying the test to the source assembly is unneeded restriction IMHO.

I'll leave it to you to decide. I just explained why I don't like using this attribute :-) it shouldn't block this PR anyway.

Reply:
> I'll leave it to you to decide.

I'll stick with the InternalsVisibleTo attribute as you haven't convinced me yet, but if you change your mind and want to resume the conversation later I'm happy to do so. I appreciate the feedback even if we haven't reached agreement on it.
I think where we are leaving off is that:
- You feel needing to touch the product assembly in response to a test change is a fairly bad place to be and merits some non-trivial effort/adding complexity to avoid it.
- My take is that I don't understand why the scenarios you described would require the test assembly name to change, and that in the unlikely case the test name did need to change I don't understand why modifying the InternalsVisibleTo attribute at the same time wouldn't be straightforward.

> did you try to run any test case under remote execution and findout if this work fine there?

If you mean have I tried to use the EventSource via an out-of-proc tool rather than via EventListener, yeah I have. I have changes to dotnet-counters in progress so I am doing some ad-hoc testing. Once the dotnet-counters work completes I expect we'll have some automated testing that does validation of the E2E scenario.

Reply:
This creates other problems.  For example, the linker won't remove anything internal from the assembly, as it doesn't know what might be accessed, and it creates problems for other analyzers that simply bail on their analysis because they can't see the whole picture.  We've avoided this at almost all costs, and haven't added it for any new code; the only projects that use it today are ones that had used it prior to moving into dotnet/runtime.

Reply:
OK, the inability to do linker analysis is a problem I grok. Is the recommendation to always use reflection to test internal surface area or is there a better solution to this? As a broader issue for the .NET ecosystem I hope we aren't making people choose between easy unit testing and good linker performance.

Reply:
Some of our projects use reflection, in particular when what needs to be accessed is minimal.  Some of our projects build (a subset of) product source into the test project(s) as well; this is common in the System.Net.* libs in particular, in their UnitTest directories.

Reply:
I compiled some of the product source into the test assembly and it was pretty straightforward, thanks for the suggestion!
InternalsVisibleTo is removed.